src/engine/renderer/Material.cpp

```diff
@@ -1778,7 +1802,56 @@ void MaterialSystem::QueueSurfaceCull( const uint32_t viewID, const frustum_t* f
 	frames[nextFrame].viewCount++;
 }
 
+void MaterialSystem::DepthReduction() {
+	if ( r_lockpvs->integer ) {
+		if ( !PVSLocked ) {
+			lockedDepthImage = frames[currentFrame].depthImage;
+		}
+
+		return;
+	}
+
+	image_t* depthImage = frames[nextFrame].depthImage;
+	int width = depthImage->width;
+	int height = depthImage->height;
+
+	gl_depthReductionShader->BindProgram( 0 );
+
+	uint32_t globalWorkgroupX = width % 8 == 0 ? width / 8 : width / 8 + 1;
+	uint32_t globalWorkgroupY = height % 8 == 0 ? height / 8 : height / 8 + 1;
+
+	GL_Bind( tr.currentDepthImage );
+	glBindImageTexture( 2, frames[nextFrame].depthImage->texnum, 0, GL_FALSE, 0, GL_WRITE_ONLY, GL_R32F );
+
+	gl_depthReductionShader->SetUniform_InitialDepthLevel( true );
+	gl_depthReductionShader->SetUniform_ViewWidth( width );
+	gl_depthReductionShader->SetUniform_ViewHeight( height );
+	gl_depthReductionShader->DispatchCompute( globalWorkgroupX, globalWorkgroupY, 1 );
+
+	for ( int i = 0; i < depthImageLevels; i++ ) {
+		width = width > 1 ? width >> 1 : 1;
+		height = height > 1 ? height >> 1 : 1;
+
+		globalWorkgroupX = width % 8 == 0 ? width / 8 : width / 8 + 1;
+		globalWorkgroupY = height % 8 == 0 ? height / 8 : height / 8 + 1;
+
+		glBindImageTexture( 1, frames[nextFrame].depthImage->texnum, i, GL_FALSE, 0, GL_READ_ONLY, GL_R32F );
+		glBindImageTexture( 2, frames[nextFrame].depthImage->texnum,i + 1, GL_FALSE, 0, GL_WRITE_ONLY, GL_R32F );
+
+		gl_depthReductionShader->SetUniform_InitialDepthLevel( false );
+		gl_depthReductionShader->SetUniform_ViewWidth( width );
+		gl_depthReductionShader->SetUniform_ViewHeight( height );
+		gl_depthReductionShader->DispatchCompute( globalWorkgroupX, globalWorkgroupY, 1 );
+
+		glMemoryBarrier( GL_SHADER_IMAGE_ACCESS_BARRIER_BIT );
```

Comment:
Is this needed before the first iteration as well?

Reply:
No.